cql3/expr/expression.cc

```diff
@@ -635,43 +611,29 @@ const auto deref = boost::adaptors::transformed([] (const managed_bytes_opt& b)
 value_list get_IN_values(
         const ::shared_ptr<term>& t, const query_options& options, const serialized_compare& comparator,
         sstring_view column_name) {
-    // RHS is prepared differently for different CQL cases.  Cast it dynamically to discern which case this is.
-    if (auto dv = dynamic_pointer_cast<lists::delayed_value>(t)) {
-        // Case `a IN (1,2,3)`.
-        const auto result_range = dv->get_elements()
-                | boost::adaptors::transformed([&] (const ::shared_ptr<term>& t) { return to_managed_bytes_opt(t->bind_and_get(options)); })
-                | non_null | deref;
-        return to_sorted_vector(std::move(result_range), comparator);
-    } else if (auto mkr = dynamic_pointer_cast<lists::marker>(t)) {
-        // Case `a IN ?`.  Collect all list-element values.
-        const auto val = mkr->bind(options);
-        if (val == constants::UNSET_VALUE) {
-            throw exceptions::invalid_request_exception(format("Invalid unset value for column {}", column_name));
-        }
-        statements::request_validations::check_not_null(val, "Invalid null value for column %s", column_name);
-        return to_sorted_vector(static_pointer_cast<lists::value>(val)->get_elements() | non_null | deref, comparator);
+    const constant in_list = evaluate_IN_list(t, options);
+    if (!in_list.is_null_or_unset() && !in_list.type->without_reversed().is_list()) {
```

Comment:
Wait, this throws if in_list is NOT null or unset?  That feels odd...

Reply:
Cases for `null` and `unset` are handled separately. In cases where this is not `null` or `unset` we have to check the type, and currently the type of a `null` constant can be `empty_type`. But you are right, it's ugly, I removed `is_null_or_unset` and moved type check below `null` and `unset` checks.